### PR TITLE
Migrate Tensor ops to pointer-based autograd broadcasting

### PR DESCRIPTION
**Pull request by wvabrinskas**
## Summary

- **Per-channel broadcasting**: Add `(W,H,D) op (1,1,D)` fast path to Tensor arithmetic operators, enabling layers to express per-channel operations (normalize, scale, bias) via standard Tensor `+`, `-`, `*`, `/` instead of manual `NumSwiftFlat` loops
- **Pointer-based `apply(axis:)` rewrite**: Replace `[Scalar]`-based `MathBlock` with `PointerMathBlock` using `TensorStorage` output/scratch buffers, eliminating intermediate array allocations in all reduction operations (sum, mean, variance, etc.)
- **Layer migrations**: Rewrite BatchNormalize, InstanceNormalize, Conv2d, TransConv2d, and DepthwiseConv2d forward/backward passes to use Tensor broadcasting instead of per-depth-slice `NumSwiftFlat` loops. BatchNormalize gamma/beta/movingMean/movingVariance converted from `[Scalar]` to `Tensor` shaped `(1,1,D)`
- **TensorMath cleanup**: Migrate `concat`, `split`, `zerosLike`, `onesLike`, and `map` to `TensorStorage` with pointer bulk copies. Remove legacy `applyAlong` fallback and `MathAlongBlock` typealias

Net result: -82 lines, fewer intermediate array allocations, and normalization/convolution layers now express math as readable Tensor expressions while retaining pointer-level performance via the broadcasting fast paths.

## Test plan

- [ ] `CI=true swift test` passes all existing tests
- [ ] Verify BatchNormalize serialization round-trip (Codable still uses `[Scalar]` arrays for `.smodel` backward compat)
- [ ] Verify InstanceNormalize gradient layout matches weights order (beta|gamma)
- [ ] Profile forward/backward pass to confirm no performance regression from Tensor broadcasting vs manual NumSwiftFlat loops

### CODE REVIEW

**Comment by wvabrinskas** [comment]
## 🤖 Neuron PR Bot Review
**PR:** #161 — Migrate Tensor ops to pointer-based autograd broadcasting
**Commit:** a1408b81

### 🐛 Bugs

- **[Critical] `movingMean`/`movingVariance` self-assignment creates reference cycles** in `BatchNormalize.apply(gradients:)`:
  ```swift
  movingMean = movingMean * momentum + batchMeansT * (1 - momentum)
  movingVariance = movingVariance * momentum + batchVariancesT * (1 - momentum)
  ```
  The `*` operator builds an autograd graph referencing the old `movingMean`, which is then assigned back to `movingMean` — exactly the anti-pattern documented in the new CLAUDE.md rule #11. Notice `gamma` and `beta` were correctly updated with `.copy()` but these two were missed. Fix: use `movingMean.copy() * momentum` (or TensorStorage-level arithmetic).

- **[Critical] `addAlong` / `subtractAlong` / `multiplyAlong` / `divideAlong` fallback paths now silently return `self` instead of computing the operation:**
  ```swift
  // e.g. in subtractAlong fallback:
  return Tensor(storage: storage.copy(), size: size, context: subtractContext(value: value))
  ```
  When input shapes don't satisfy `_broadcastAlongFastPath` conditions, the result is just a copy of `self` with a context attached — `value` is never subtracted. The previous `applyAlong` fallback actually performed the math. Any caller hitting a shape that misses both fast paths gets silently wrong outputs, which could corrupt training without any error.

- **[Medium] `BatchNormalize.normalize` captures a Tensor with autograd graph inside `TensorContext`:** The `normalized` tensor returned from `(inputs - meanT) / stdT` carries its own autograd graph. It's then captured in the outer `TensorContext` closure. This creates a closure → Tensor → autograd graph → inputs chain that will be retained for the entire backward pass, inflating memory usage during batch training. The old code stored plain `TensorStorage` snapshots to avoid this. Consider `.detach()` or storing `normalization.normalized.storage` (strip autograd before capture).

- **[Low] `DepthwiseConv2d` bias addition introduces an autograd `+` node inside what should be a raw forward computation helper:** The new path `return Tensor(storage: resultStorage, size: outputSize) + biasT` returns a Tensor with an `addContext` attached. If the calling `forward` wraps this in its own `TensorContext`, that context node is dead weight that never participates in backprop. `Conv2d` correctly strips this by returning `result.storage` — `DepthwiseConv2d` should do the same for consistency.

### 🔄 Code Consistency

- **`BatchNormalize.apply(gradients:)` inconsistently uses `.copy()` for gamma/beta but not for movingMean/movingVariance.** Since the repo now has an explicit rule against Tensor self-assignment, all four should follow the same safe pattern.

- **`_broadcastPerChannelFastPath` is private to `TensorMath.swift` but essentially duplicates the structure of `_broadcastAlongFastPath`.** Consider merging them or at minimum naming them symmetrically. The new method name `_broadcastPerChannelFastPath` is fine but doesn't follow a consistent naming scheme with the existing `_broadcastAlongFastPath`.

- **`Normalization` struct changed `normalized` from `TensorStorage` to `Tensor` and `std` to `stds: [Tensor.Scalar]`** — a reasonable change, but the mixed approach (Tensor for `normalized`, scalar array for `stds`) is inconsistent. Either make it all Tensor-based or all scalar-based.

- **`PointerMathBlock` replaces both `MathBlock` and `MathAlongBlock`** but now `apply(axis:)` blocks receive a raw `UnsafeMutablePointer` with no bounds safety. If any caller passes an incorrect `count`, memory reads are silent UB. A lightweight bounds-checked abstraction or at least a precondition comment would help.

### ⚡ Performance

- **`BatchNormalize.normalize` now creates intermediate Tensors for `meanT`, `stdT`, `normalized`, `output`** — each allocation is heap-backed `TensorStorage`. The old code used two reusable scratch buffers and a single output buffer. For small depth values (typical BN use), the new code trades simplicity for extra allocations per forward pass. Not critical, but worth noting.

- **`InstanceNormalize.normalize` and `.backward` now recompute means/stds in separate loops** where the old code did everything in one pass per depth slice. The Tensor-level ops are cleaner, but for large spatial dims the two-pass approach (one for means, one for stds) doubles memory reads on the input. Could be unified with a combined mean+variance pointer pass using `NumSwiftFlat`.

- **`_broadcastPerChannelFastPath` is an excellent addition** — the per-channel scalar broadcast is a very common pattern in norm layers and the `NumSwiftFlat` pointer loop will be dramatically faster than the previous element-wise storage arithmetic. Good call.

- **`split`, `concat`, and `map` now use pointer `update(from:count:)` instead of element-wise Swift array copies** — solid improvement throughout.

### ✅ Summary

This is a well-motivated PR that correctly identifies and resolves the Tensor self-assignment reference cycle problem in the autograd graph, and the pointer-based broadcasting fast path is a real performance win. However, there are two correctness issues that need fixing before merge: the `movingMean`/`movingVariance` self-assignment in `apply(gradients:)` has the exact bug this PR set out to fix, and the `applyAlong` fallback removal silently breaks any broadcasting path that misses the fast-path shape checks.
